h2o-core/src/main/java/water/ClientHeartBeatCheckThread.java

```diff
@@ -0,0 +1,36 @@
+package water;
+
+import water.util.Log;
+
+/**
+ * This thread checks whether the heartbeat from the bully client has been seen before the timeout and if not, it kills
+ * the whole h2o cloud
+ */
+public class ClientHeartBeatCheckThread extends Thread {
+    public ClientHeartBeatCheckThread() {
+        super("ClientHeartbeatCheckThread");
+        setDaemon(true);
+    }
+
+    @Override
+    public void run(){
+        while(true) {
+            for (H2ONode n : H2O.getFlatfile()) {
+                if (n._heartbeat._bully_client) {
+                    if(n._last_heard_from + H2O.ARGS.bully_client_timeout >= System.currentTimeMillis()) {
+                        Log.warn("Stopping H2O cloud since bully client hasn't seen heartbeat in specified timeout");
```

Comment:
good point. I have whole new version of this change locally, will push it tonight with this update